test_util/cluster_api.py

```diff
@@ -263,39 +276,56 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
 
         assert len(self.masters) == len(self.public_masters)
 
-        # URI must include scheme
-        assert dcos_uri.startswith('http')
-        parse_result = urlparse(dcos_uri)
-        self.scheme = parse_result.scheme
-        self.dns_host = parse_result.netloc.split(':')[0]
-
-        # Make URI never end with /
-        self.dcos_uri = dcos_uri.rstrip('/')
-
     def get_user_session(self, user):
         """ Return a copy of self instead of new instance of ClusterApi
         so that children of this class return siblings, not parents
         """
         new_session = copy.deepcopy(self)
         if user:
             user.authenticate(self)
+            new_session.default_headers = user.auth_header
+        else:
+            new_session.default_headers = {}
         new_session.web_auth_default_user = user
         return new_session
 
+    def get_node_url(self, node, port=None):
+        """
+        Args:
+            node: (str) the hostname of the node to be requested from, if node=None,
+                then public cluster address (see environment DCOS_DNS_ADDRESS)
+            port: (int) port to be requested at. If port=None, the default port
+                for that given node type will be used
+        Returns:
+            fully-qualified URL string for this API
+        """
+        if node in self.masters:
+            role = 'master'
+        elif node in self.all_slaves:
+            role = 'agent'
+        else:
+            raise Exception('Node {} is not recognized within the DC/OS cluster'.format(node))
+
+        if port is None:
+            port = ADMINROUTER_PORT_MAPPING[role][self.scheme]
+
+        # do not explicitly declare defaults
+        if (port == 80 and self.scheme == 'http') or (port == 443 and self.scheme == 'https'):
+            netloc = node
+        else:
+            netloc = '{}:{}'.format(node, port)
+
+        return '{}://{}'.format(self.scheme, netloc)
+
     def get_client(self, path, default_headers=None):
-        return test_util.helpers.ApiClient(
+        return test_util.helpers.ApiClient.from_cluster(
             cluster=self,
-            user=self.web_auth_default_user,
             api_base=path,
-            ca_cert_path=self.ca_cert_path,
-            default_headers=default_headers)
+            default_headers=None)
```

Comment:
why is this not passing through `self.default_headers if default_headers is None else default_headers`?

Lets people set them to empty if they want / need, provides the same responses they would have gotten by hand-prefixing their requests.